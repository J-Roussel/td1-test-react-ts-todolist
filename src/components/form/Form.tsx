import React, {FC} from "react";
import './form.css';
import { IForm } from "../../utils/interfaces";


const Form:FC<IForm> = ({valueName, valueDescri, initialValue, handleChange, addTask}):JSX.Element => {

    
    return (
        <div className='form-container'>
            <div className="form">
                <div className="info">
                    <div className="name" >
                        <label className="label-name" htmlFor="name">&nbsp; &nbsp; &nbsp;Name</label>
                        <input className="input-name" type="text" value={valueName}  onChange={handleChange} name='inputName' placeholder={initialValue[0]}/>
                    </div>
                    <div className="description">
                        <label className="label-description" htmlFor="description">&nbsp; &nbsp; &nbsp;Description</label>
                        <input className="input-description" type="textarea" value={valueDescri} onChange={handleChange} name='inputDescri' placeholder={initialValue[1]}/>
                    </div>
                </div>
                <input type="button" className="valider" value="ADD ToDo" onClick={addTask}/>
            </div>        
        </div>
    );
}


export default Form;