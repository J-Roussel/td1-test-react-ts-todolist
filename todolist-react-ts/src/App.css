*{
    margin: 0;
    padding: 0;
    background-color: #0B2B40;
  }
  
  
  h1{
    text-align: center;
    margin-top: 2%;
    margin-bottom: 10px;
    color: #F2BE22;
  }


.card {
  width: 100%;
  min-height: 550px;
  margin-top: 20px;
  display: flex;
  justify-content: space-evenly;
}

#left-card-title, #left-task{
    background-color: #F2BE22;
}

#right-card-title, #right-task{  
    background-color: #267365;
}

.task {
  width: 98%;
  margin: auto;
  margin-top: 2%;
  min-height: 90px;
  border-bottom-left-radius: 5%;
  border-bottom-right-radius: 5%;
}

.action-btn{
  margin-top: 4%;
  margin-left: 55%;
  display: flex;
  justify-content: space-evenly;
  width: 40%;
}

.completed-btn {
  width: 45%;
  background-color: #185359;
  color: #F2BE22;
  border: 1px solid #F2BE22;
}

.completed-btn:hover {
  background-color: #267365;
}

.remove-btn {
  width: 45%;
  background-color: #8C1F28;
  color: white;
  border: 1px solid white;
}

.remove-btn:hover {
  background-color: #D92525;
}