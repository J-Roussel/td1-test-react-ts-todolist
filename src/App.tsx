import React, {useState} from 'react';
import './App.css';
import Form from './components/form/Form';
import Card from './components/card/card';
import { ITask } from "./utils/interfaces";


function App() {


  //Default values of the the inputs of the form
  let initialValue: string[] = ["Name of the task...", "Description of the task..."];
  const [valueName, setValueName] = useState<string>("");
  const [valueDescri, setValueDescri] = useState<string>("");

  //useState for the ToDoList
  const [toDoList, setToDo] = useState<ITask[]>([]);

  const [statut, setStatut] = useState<boolean>(false);

  // mark as done
  const asDoneBtn = ():void => {
    setStatut(!statut);
  }

  //onChange function
  const handleChange = (e:React.ChangeEvent<HTMLInputElement>):void => {
    (e.target.name==='inputName')? setValueName(e.currentTarget.value) : setValueDescri(e.currentTarget.value);
  };  

  //function that allows to add the task given by user in the ToDo card
  const addTask = ():void => {
    const newTask = {name:valueName, description:valueDescri};
    setToDo([...toDoList, newTask]);
    setStatut(false);
    setValueName("");
    setValueDescri("");
  }

  //function that allows to remove the task from its card
  const removeTask = (taskNameToDelete:string):void => {
    setToDo(toDoList.filter((task:ITask) => {
      return (task.name !== taskNameToDelete);
    }));
  }

  return (
    <React.Fragment>
      <div className='container'>
        <h1>ToDo List React Typescript</h1>
        <Form valueName={valueName} 
              valueDescri={valueDescri} 
              initialValue={initialValue}
              handleChange={handleChange}
              addTask={addTask}
        />

        <div className='card'>
          <div className="left-card">
            <Card titre="ToDo" id="left-card-title">
              <>
                {toDoList.map((task:ITask) => {
                  if((task.name !== "" || task.description !== "") && statut === false){
                    return (
                      <div className="task">
                        <div className="task-data" >
                          <h4>{task.name}</h4>
                          <p className="taskDescription">{task.description}</p>
                        </div>
                        <div className="action-btn">
                              <input className="completed-btn" type="button" value="completed" onClick={asDoneBtn}/>
                              <input className="remove-btn"  type="button" value="remove"onClick={()=>{removeTask(task.name)}}/>
                        </div>
                      </div>                
                    );
                  }
                  return "";
                })}
              </>
            </Card>
          </div>

          <div className=''>
            <Card titre='Done' id='right-card-title'>
              <>
                {toDoList.map((task:ITask) => {
                      if(statut === true){
                        return (
                          <div className="task">
                            <div className="task-data" >
                              <h4>{task.name}</h4>
                              <p className="taskDescription">{task.description}</p>
                            </div>
                            <div className="action-btn">
                                  <input className="remove-btn"  type="button" value="remove"onClick={()=>{removeTask(task.name)}}/>
                            </div>
                          </div>                
                        );
                      }
                    })}
                
                </>
              </Card>
          </div>

        </div>

      </div>
    </React.Fragment>
  );
}

export default App;
