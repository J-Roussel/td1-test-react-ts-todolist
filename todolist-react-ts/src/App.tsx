import React from 'react';
import './App.css';
import Form from './components/form/Form';

function App() {
  return (
    <React.Fragment>
      <div className='container'>
        <h1>ToDo List React Typescript</h1>
        <Form />

      </div>
    </React.Fragment>
  );
}

export default App;
