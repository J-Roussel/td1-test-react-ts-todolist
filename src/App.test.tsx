import renderer from "react-test-renderer";
import App from './App';
import Form from './components/form/Form';
import Card  from './components/card/card';


describe('Unit test for App component', () => {
  test('renders app component correctly', () => {
    const APPCOMPONENT = renderer.create(<App/>);
    let tree = APPCOMPONENT.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('render the correct title and subtitle', ()=>{
    const APPCOMPONENT = renderer.create(<App/>);
    const TEXTELEMENT = APPCOMPONENT.root.findByType("h1").children;
    console.log(TEXTELEMENT.toString());
    
    expect(TEXTELEMENT.toString()).toMatch(/ToDo List React Typescript/i);
  });

  test('all components renders normally', () => {
    const APPCOMPONENT = renderer.create(<App/>);
    const TEXTELEMENT = APPCOMPONENT.root;

    let formComponent = TEXTELEMENT.findByType(Form);
    expect(formComponent).toBeDefined();

    let cardComponent = TEXTELEMENT.findAllByType(Card);
    console.log(cardComponent);
    
    expect(cardComponent.length).toBe(2);

  });


})
