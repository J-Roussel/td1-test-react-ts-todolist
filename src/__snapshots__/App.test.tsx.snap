// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test unitaire renders app component correctly 1`] = `
<div
  className="container"
>
  <h1>
    ToDo List React Typescript
  </h1>
  <div
    className="form-container"
  >
    <div
      className="form"
    >
      <div
        className="info"
      >
        <div
          className="name"
        >
          <label
            className="label-name"
            htmlFor="name"
          >
                 Name
          </label>
          <input
            className="input-name"
            name="inputName"
            onChange={[Function]}
            placeholder="Name of the task..."
            type="text"
            value=""
          />
        </div>
        <div
          className="description"
        >
          <label
            className="label-description"
            htmlFor="description"
          >
                 Description
          </label>
          <input
            className="input-description"
            name="inputDescri"
            onChange={[Function]}
            placeholder="Description of the task..."
            type="textarea"
            value=""
          />
        </div>
      </div>
      <input
        className="valider"
        onClick={[Function]}
        type="button"
        value="ADD ToDo"
      />
    </div>
  </div>
  <div
    className="card"
  >
    <div
      className="left-card"
    >
      <div
        className="card-container"
      >
        <h3
          className="title"
          id="left-card-title"
        >
          ToDo
        </h3>
        <p
          className="add-description"
        >
          Add 
          <span>
            +
          </span>
        </p>
      </div>
    </div>
    <div
      className=""
    >
      <div
        className="card-container"
      >
        <h3
          className="title"
          id="right-card-title"
        >
          Done
        </h3>
        <p
          className="add-description"
        >
          Add 
          <span>
            +
          </span>
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`Unit test for App component renders app component correctly 1`] = `
<div
  className="container"
>
  <h1>
    ToDo List React Typescript
  </h1>
  <div
    className="form-container"
  >
    <div
      className="form"
    >
      <div
        className="info"
      >
        <div
          className="name"
        >
          <label
            className="label-name"
            htmlFor="name"
          >
                 Name
          </label>
          <input
            className="input-name"
            name="inputName"
            onChange={[Function]}
            placeholder="Name of the task..."
            type="text"
            value=""
          />
        </div>
        <div
          className="description"
        >
          <label
            className="label-description"
            htmlFor="description"
          >
                 Description
          </label>
          <input
            className="input-description"
            name="inputDescri"
            onChange={[Function]}
            placeholder="Description of the task..."
            type="textarea"
            value=""
          />
        </div>
      </div>
      <input
        className="valider"
        onClick={[Function]}
        type="button"
        value="ADD ToDo"
      />
    </div>
  </div>
  <div
    className="card"
  >
    <div
      className="left-card"
    >
      <div
        className="card-container"
      >
        <h3
          className="title"
          id="left-card-title"
        >
          ToDo
        </h3>
        <p
          className="add-description"
        >
          Add 
          <span>
            +
          </span>
        </p>
      </div>
    </div>
    <div
      className=""
    >
      <div
        className="card-container"
      >
        <h3
          className="title"
          id="right-card-title"
        >
          Done
        </h3>
        <p
          className="add-description"
        >
          Add 
          <span>
            +
          </span>
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders app component  1`] = `
<div
  className="container"
>
  <h1>
    ToDo List React Typescript
  </h1>
  <div
    className="form-container"
  >
    <div
      className="form"
    >
      <div
        className="info"
      >
        <div
          className="name"
        >
          <label
            className="label-name"
            htmlFor="name"
          >
                 Name
          </label>
          <input
            className="input-name"
            name="inputName"
            onChange={[Function]}
            placeholder="Name of the task..."
            type="text"
            value=""
          />
        </div>
        <div
          className="description"
        >
          <label
            className="label-description"
            htmlFor="description"
          >
                 Description
          </label>
          <input
            className="input-description"
            name="inputDescri"
            onChange={[Function]}
            placeholder="Description of the task..."
            type="textarea"
            value=""
          />
        </div>
      </div>
      <input
        className="valider"
        onClick={[Function]}
        type="button"
        value="ADD ToDo"
      />
    </div>
  </div>
  <div
    className="card"
  >
    <div
      className="left-card"
    >
      <div
        className="card-container"
      >
        <h3
          className="title"
          id="left-card-title"
        >
          ToDo
        </h3>
        <p
          className="add-description"
        >
          Add 
          <span>
            +
          </span>
        </p>
      </div>
    </div>
    <div
      className=""
    >
      <div
        className="card-container"
      >
        <h3
          className="title"
          id="right-card-title"
        >
          Done
        </h3>
        <p
          className="add-description"
        >
          Add 
          <span>
            +
          </span>
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders correctly 1`] = `
<div
  className="container"
>
  <h1>
    ToDo List React Typescript
  </h1>
  <div
    className="form-container"
  >
    <div
      className="form"
    >
      <div
        className="info"
      >
        <div
          className="name"
        >
          <label
            className="label-name"
            htmlFor="name"
          >
                 Name
          </label>
          <input
            className="input-name"
            name="inputName"
            onChange={[Function]}
            placeholder="Name of the task..."
            type="text"
            value=""
          />
        </div>
        <div
          className="description"
        >
          <label
            className="label-description"
            htmlFor="description"
          >
                 Description
          </label>
          <input
            className="input-description"
            name="inputDescri"
            onChange={[Function]}
            placeholder="Description of the task..."
            type="textarea"
            value=""
          />
        </div>
      </div>
      <input
        className="valider"
        onClick={[Function]}
        type="button"
        value="ADD ToDo"
      />
    </div>
  </div>
  <div
    className="card"
  >
    <div
      className="left-card"
    >
      <div
        className="card-container"
      >
        <h3
          className="title"
          id="left-card-title"
        >
          ToDo
        </h3>
        <p
          className="add-description"
        >
          Add 
          <span>
            +
          </span>
        </p>
      </div>
    </div>
    <div
      className=""
    >
      <div
        className="card-container"
      >
        <h3
          className="title"
          id="right-card-title"
        >
          Done
        </h3>
        <p
          className="add-description"
        >
          Add 
          <span>
            +
          </span>
        </p>
      </div>
    </div>
  </div>
</div>
`;
